modules/gdscript/gdscript_compiler.cpp

```diff
@@ -1056,33 +1056,61 @@ GDScriptCodeGenerator::Address GDScriptCompiler::_parse_expression(CodeGen &code
 
 				// Set back the values into their bases.
 				for (const ChainInfo &info : set_chain) {
-					if (!info.is_named) {
-						gen->write_set(info.base, info.key, assigned);
-						if (info.key.mode == GDScriptCodeGenerator::Address::TEMPORARY) {
-							gen->pop_temporary();
+					bool known_type = assigned.type.has_type;
+					bool is_shared = Variant::is_type_shared(assigned.type.builtin_type);
+
+					if (!known_type) {
+						// Jump shared values since they are already updated in-place.
+						gen->write_jump_if_shared(assigned);
+					}
+					if (known_type && !is_shared) {
+						if (!info.is_named) {
+							gen->write_set(info.base, info.key, assigned);
+							if (info.key.mode == GDScriptCodeGenerator::Address::TEMPORARY) {
+								gen->pop_temporary();
+							}
+						} else {
+							gen->write_set_named(info.base, info.name, assigned);
 						}
-					} else {
-						gen->write_set_named(info.base, info.name, assigned);
+					}
+					if (!known_type) {
+						gen->write_end_jump_if_shared();
 					}
```

Comment:
The above probably needs a similar logic to the part outside the loop (`if (!known_type || !is_shared)`) because currently `jump_if_shared` only jumps to the very next instruction :slightly_smiling_face: 

```gdscript
extends Node3D

func _ready():
    var dict : Dictionary = {}
    dict["A"]["B"].origin = Vector2()
```
```
Disassembling _ready()
 0: line 4:     var dict : Dictionary = {}
 2: make_dict stack(4) = {}
 5: assign stack(3) = stack(4)
 8: line 5:     dict["A"]["B"].origin = Vector2()
 10: get keyed validated stack(5) = stack(3)[const("A")]
 15: get keyed stack(6) = stack(5)[const("B")]
 19: set_named stack(6)["origin"] = const((0, 0))
 23: jump-if-shared stack(6) to 26
 26: jump-if-shared stack(5) to 29
 29: == END ==
```